test/with_dummyserver/test_connectionpool.py

```diff
@@ -870,5 +870,65 @@ def test_redirect_after(self):
         self.assertEqual(r.status, 200)
         self.assertTrue(delta < 1)
 
+class TestRetryWithTimeout(HTTPDummyServerTestCase):
+    def setUp(self):
+        self.pool = HTTPConnectionPool(self.host, self.port, timeout=0.1)
+
+    def test_retries_put_filehandle(self):
+        """HTTP PUT retry with a file-like object should not timeout"""
+        retry = Retry(total=3, status_forcelist=[418])
+        # httplib reads in 8k chunks; use a larger content length
+        content_length = 65535
+        uploaded_file = io.BytesIO(b'A' * content_length)
+        headers = {'test-name': 'test_put_fileobj',
+                   'Content-Length': str(content_length)}
+        resp = self.pool.urlopen('PUT', '/successful_retry',
+                                 headers=headers,
+                                 retries=retry,
+                                 body=uploaded_file,
+                                 assert_same_host=False, redirect=False)
+        self.assertEqual(resp.status, 200)
+
+    def test_redirect_put_file(self):
+        """PUT with file object should work with a redirection response"""
+        retry = Retry(total=3, status_forcelist=[418])
+        # httplib reads in 8k chunks; use a larger content length
+        content_length = 65535
+        uploaded_file = io.BytesIO(b'A' * content_length)
+        headers = {'test-name': 'test_redirect_put_fileobj_timeout',
+                   'Content-Length': str(content_length)}
+        resp = self.pool.urlopen('PUT', '/redirect?target=/successful_retry',
+                                 headers=headers,
+                                 retries=retry,
+                                 body=uploaded_file,
+                                 assert_same_host=False, redirect=True)
+        self.assertEqual(resp.status, 200)
```

Comment:
I'm actually not sure how we'd get that information with the current testing structure. The dummyserver doesn't return our read request data and we can't access the request from the response object.

I've tried sending the redirect to /put but test aborts on ```TypeError: put() takes exactly 1 argument (2 given)``` because Tornado is passing more params than we're expecting. I can't find any tests calling '/put' or '/post' so I'm not sure they've ever been used?

I've also tried '/echo' which would normally be what we want but it just passes us back an empty string. I thought we possibly weren't rewinding, but even passing a file-like object directly to '/echo' without a redirect will still return ''. It looks like Tornado is consuming the file objects and then not passing that data on to the handler.